File: tests/go/vectors.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

const vectorsDir = "../../test-vectors"

// AddressVector represents an address parsing test vector
type AddressVector struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Input       string `json:"input"`
	Expected    struct {
		FullHex     string `json:"full_hex"`
		ShortString string `json:"short_string"`
		BytesHex    string `json:"bytes_hex"`
		LastByte    int    `json:"last_byte"`
	} `json:"expected"`
}

// MnemonicVector represents a mnemonic derivation test vector
type MnemonicVector struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Input       struct {
		Mnemonic       string `json:"mnemonic"`
		Passphrase     string `json:"passphrase"`
		DerivationPath string `json:"derivation_path"`
	} `json:"input"`
	Expected struct {
		SeedHex       string `json:"seed_hex"`
		PrivateKeyHex string `json:"private_key_hex"`
		PublicKeyHex  string `json:"public_key_hex"`
		AuthKeyHex    string `json:"auth_key_hex"`
		Address       string `json:"address"`
	} `json:"expected"`
}

// HashVector represents a hash test vector
type HashVector struct {
	Name        string `json:"name"`
	Input       string `json:"input"`
	InputHex    string `json:"input_hex"`
	ExpectedHex string `json:"expected_hex"`
}

// BcsVector represents a BCS encoding test vector
type BcsVector struct {
	Name   string      `json:"name"`
	Type   string      `json:"type"`
	Value  interface{} `json:"value"`
	BcsHex string      `json:"bcs_hex"`
}

// loadVectorFile loads a JSON test vector file
func loadVectorFile(filename string) (map[string]interface{}, error) {
	path := filepath.Join(vectorsDir, filename)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var result map[string]interface{}
	err = json.Unmarshal(data, &result)
	return result, err
}

// GetAddressVectors loads address test vectors
func GetAddressVectors() (map[string]interface{}, error) {
	return loadVectorFile("addresses.json")
}

// GetAddressParsingVectors returns address parsing vectors
func GetAddressParsingVectors() ([]AddressVector, error) {
	data, err := GetAddressVectors()
	if err != nil {
		return nil, err
	}

	vectorsRaw, ok := data["parsing_vectors"].([]interface{})
	if !ok {
		return nil, nil
	}

	var vectors []AddressVector
	for _, v := range vectorsRaw {
		jsonBytes, _ := json.Marshal(v)
		var av AddressVector
		json.Unmarshal(jsonBytes, &av)
		vectors = append(vectors, av)
	}
	return vectors, nil
}

// GetMnemonicVectors loads mnemonic test vectors
func GetMnemonicVectors() (map[string]interface{}, error) {
	return loadVectorFile("mnemonics.json")
}

// GetEd25519DerivationVectors returns Ed25519 derivation vectors
func GetEd25519DerivationVectors() ([]MnemonicVector, error) {
	data, err := GetMnemonicVectors()
	if err != nil {
		return nil, err
	}

	vectorsRaw, ok := data["ed25519_derivation_vectors"].([]interface{})
	if !ok {
		return nil, nil
	}

	var vectors []MnemonicVector
	for _, v := range vectorsRaw {
		jsonBytes, _ := json.Marshal(v)
		var mv MnemonicVector
		json.Unmarshal(jsonBytes, &mv)
		vectors = append(vectors, mv)
	}
	return vectors, nil
}

// GetSignatureVectors loads signature test vectors
func GetSignatureVectors() (map[string]interface{}, error) {
	return loadVectorFile("signatures.json")
}

// GetSha3256Vectors returns SHA3-256 hash vectors
func GetSha3256Vectors() ([]HashVector, error) {
	data, err := GetSignatureVectors()
	if err != nil {
		return nil, err
	}

	hashing, ok := data["hashing"].(map[string]interface{})
	if !ok {
		return nil, nil
	}

	vectorsRaw, ok := hashing["sha3_256"].([]interface{})
	if !ok {
		return nil, nil
	}

	var vectors []HashVector
	for _, v := range vectorsRaw {
		jsonBytes, _ := json.Marshal(v)
		var hv HashVector
		json.Unmarshal(jsonBytes, &hv)
		vectors = append(vectors, hv)
	}
	return vectors, nil
}

// GetBcsVectors loads BCS encoding test vectors
func GetBcsVectors() (map[string]interface{}, error) {
	return loadVectorFile("bcs.json")
}

// HexToBytes converts a hex string to bytes
func HexToBytes(hexStr string) ([]byte, error) {
	cleanHex := strings.TrimPrefix(hexStr, "0x")
	return hex.DecodeString(cleanHex)
}

// BytesToHex converts bytes to a hex string with 0x prefix
func BytesToHex(data []byte) string {
	return "0x" + hex.EncodeToString(data)
}

File: tests/rust/src/support/vectors.rs
//! Test vector loading utilities.
//!
//! Loads deterministic test vectors from the shared JSON files.

use serde::Deserialize;
use std::fs;
use std::path::PathBuf;

/// Get the path to the test vectors directory.
fn vectors_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join("test-vectors")
}

// =============================================================================
// Address Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct AddressVectors {
    pub version: String,
    pub description: String,
    pub parsing_vectors: Vec<AddressParsingVector>,
    pub constants: Vec<AddressConstant>,
    pub invalid_inputs: Vec<InvalidAddressInput>,
    pub bcs_serialization: Vec<BcsSerializationVector>,
}

#[derive(Debug, Deserialize)]
pub struct AddressParsingVector {
    pub name: String,
    pub description: String,
    pub input: String,
    pub expected: AddressExpected,
}

#[derive(Debug, Deserialize)]
pub struct AddressExpected {
    pub full_hex: String,
    pub short_string: String,
    #[serde(default)]
    pub bytes_hex: Option<String>,
    #[serde(default)]
    pub last_byte: Option<u8>,
    #[serde(default)]
    pub last_two_bytes_hex: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AddressConstant {
    pub name: String,
    pub description: String,
    pub expected: AddressConstantExpected,
}

#[derive(Debug, Deserialize)]
pub struct AddressConstantExpected {
    pub full_hex: String,
    pub short_string: String,
    #[serde(default)]
    pub bytes: Option<Vec<u8>>,
}

#[derive(Debug, Deserialize)]
pub struct InvalidAddressInput {
    pub name: String,
    pub input: String,
    pub expected_error: Option<String>,
    pub reason: String,
    #[serde(default)]
    pub note: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct BcsSerializationVector {
    pub name: String,
    pub description: String,
    pub input: String,
    pub expected: BcsExpected,
}

#[derive(Debug, Deserialize)]
pub struct BcsExpected {
    pub bcs_hex: String,
    pub length: usize,
}

impl AddressVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("addresses.json");
        let content = fs::read_to_string(&path)?;
        let vectors: AddressVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// Signature Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct SignatureVectors {
    pub version: String,
    pub ed25519: Vec<Ed25519TestVector>,
    #[serde(default)]
    pub secp256k1: Vec<Secp256k1TestVector>,
}

#[derive(Debug, Deserialize)]
pub struct Ed25519TestVector {
    pub name: String,
    pub description: String,
    pub private_key_hex: String,
    pub public_key_hex: String,
    #[serde(default)]
    pub address_hex: Option<String>,
    #[serde(default)]
    pub message_hex: Option<String>,
    #[serde(default)]
    pub signature_hex: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Secp256k1TestVector {
    pub name: String,
    pub description: String,
    pub private_key_hex: String,
    pub public_key_hex: String,
    #[serde(default)]
    pub address_hex: Option<String>,
    #[serde(default)]
    pub message_hex: Option<String>,
    #[serde(default)]
    pub signature_hex: Option<String>,
}

impl SignatureVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("signatures.json");
        let content = fs::read_to_string(&path)?;
        let vectors: SignatureVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// Mnemonic Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct MnemonicVectors {
    pub version: String,
    pub derivation_vectors: Vec<MnemonicDerivationVector>,
}

#[derive(Debug, Deserialize)]
pub struct MnemonicDerivationVector {
    pub name: String,
    pub description: String,
    pub mnemonic: String,
    pub passphrase: String,
    pub derivation_path: String,
    pub expected: MnemonicExpected,
}

#[derive(Debug, Deserialize)]
pub struct MnemonicExpected {
    pub private_key_hex: String,
    pub public_key_hex: String,
    pub address_hex: String,
}

impl MnemonicVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("mnemonics.json");
        let content = fs::read_to_string(&path)?;
        let vectors: MnemonicVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// Type Tag Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct TypeTagVectors {
    pub version: String,
    pub primitives: Vec<TypeTagVector>,
    pub vectors: Vec<TypeTagVector>,
    pub structs: Vec<TypeTagVector>,
    pub invalid: Vec<InvalidTypeTagVector>,
}

#[derive(Debug, Deserialize)]
pub struct TypeTagVector {
    pub name: String,
    pub input: String,
    pub expected: TypeTagExpected,
}

#[derive(Debug, Deserialize)]
pub struct TypeTagExpected {
    pub canonical: String,
    #[serde(default)]
    pub bcs_hex: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct InvalidTypeTagVector {
    pub name: String,
    pub input: String,
    pub reason: String,
}

impl TypeTagVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("type-tags.json");
        let content = fs::read_to_string(&path)?;
        let vectors: TypeTagVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// BCS Serialization Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct BcsVectors {
    pub version: String,
    pub primitives: Vec<BcsPrimitiveVector>,
    #[serde(default)]
    pub strings: Vec<BcsStringVector>,
    #[serde(default)]
    pub sequences: Vec<BcsSequenceVector>,
}

#[derive(Debug, Deserialize)]
pub struct BcsPrimitiveVector {
    pub name: String,
    #[serde(rename = "type")]
    pub type_name: String,
    pub value: serde_json::Value,
    pub expected_hex: String,
}

#[derive(Debug, Deserialize)]
pub struct BcsStringVector {
    pub name: String,
    pub value: String,
    pub expected_hex: String,
}

#[derive(Debug, Deserialize)]
pub struct BcsSequenceVector {
    pub name: String,
    #[serde(rename = "type")]
    pub type_name: String,
    pub value: Vec<serde_json::Value>,
    pub expected_hex: String,
}

impl BcsVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("bcs.json");
        let content = fs::read_to_string(&path)?;
        let vectors: BcsVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// Transaction Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct TransactionVectors {
    pub version: String,
    #[serde(default)]
    pub raw_transactions: Vec<RawTransactionVector>,
    #[serde(default)]
    pub signed_transactions: Vec<SignedTransactionVector>,
}

#[derive(Debug, Deserialize)]
pub struct RawTransactionVector {
    pub name: String,
    pub description: String,
    pub fields: RawTransactionFields,
    pub expected: TransactionExpected,
}

#[derive(Debug, Deserialize)]
pub struct RawTransactionFields {
    pub sender: String,
    pub sequence_number: u64,
    pub max_gas_amount: u64,
    pub gas_unit_price: u64,
    pub expiration_timestamp_secs: u64,
    pub chain_id: u8,
    pub payload_type: String,
    #[serde(default)]
    pub payload_module: Option<String>,
    #[serde(default)]
    pub payload_function: Option<String>,
    #[serde(default)]
    pub payload_type_args: Option<Vec<String>>,
    #[serde(default)]
    pub payload_args: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct TransactionExpected {
    #[serde(default)]
    pub bcs_hex: Option<String>,
    #[serde(default)]
    pub signing_message_hex: Option<String>,
    #[serde(default)]
    pub hash: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SignedTransactionVector {
    pub name: String,
    pub description: String,
    pub private_key_hex: String,
    pub raw_transaction_bcs: String,
    pub expected: SignedTransactionExpected,
}

#[derive(Debug, Deserialize)]
pub struct SignedTransactionExpected {
    pub bcs_hex: String,
    pub hash: String,
}

impl TransactionVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("transactions.json");
        let content = fs::read_to_string(&path)?;
        let vectors: TransactionVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}

// =============================================================================
// Multi-Signature Test Vectors
// =============================================================================

#[derive(Debug, Deserialize)]
pub struct MultiSigVectors {
    pub version: String,
    #[serde(default)]
    pub multi_ed25519: Vec<MultiEd25519Vector>,
    #[serde(default)]
    pub multi_key: Vec<MultiKeyVector>,
}

#[derive(Debug, Deserialize)]
pub struct MultiEd25519Vector {
    pub name: String,
    pub description: String,
    pub public_keys_hex: Vec<String>,
    pub threshold: u8,
    pub expected: MultiSigExpected,
}

#[derive(Debug, Deserialize)]
pub struct MultiKeyVector {
    pub name: String,
    pub description: String,
    pub keys: Vec<MultiKeyEntry>,
    pub threshold: u8,
    pub expected: MultiSigExpected,
}

#[derive(Debug, Deserialize)]
pub struct MultiKeyEntry {
    pub key_type: String,
    pub public_key_hex: String,
}

#[derive(Debug, Deserialize)]
pub struct MultiSigExpected {
    pub public_key_bcs_hex: String,
    pub address_hex: String,
}

impl MultiSigVectors {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        let path = vectors_dir().join("multi-sig.json");
        let content = fs::read_to_string(&path)?;
        let vectors: MultiSigVectors = serde_json::from_str(&content)?;
        Ok(vectors)
    }
}
